"Enterprise Fee Summary" report

kristinalim opened the pull request:
#### What? Why?

~Closes~ Relates to #2616 
~Closes~ Relates to #2656 
~Closes~ Relates to #2672 

#### What should we test?

This is best to be tested with existing OCs (if already available) that would already have data for this report. Before testing, it would be helpful to identify hubs, producers, and coordinators that would have data needed for Step 2 below.

Each of the following roles would be checked for Steps 2 and 3.

* Admin
* Hub
* Producer
* Coordinator

To test:

1. Check that HTML and CSV report generation work. These are using the same data objects, so except occasionally, for Steps 2 and 3 there is no need to check results for both formats - just checking one should be sufficient.
2. For each user role listed above, check that these data are properly summarized:
    a. Coordinator fee for order cycle
    b. Coordinator fee for incoming exchange
    c. Coordinator fee for outgoing exchange
    d. Producer fee for incoming exchange
    e. Distributor fee for outgoing exchange 
    f. Shipping fee
    g. Payment method fee
3. Do a smoke test that the correct filter options are available to each user role.
4. Test that using each filter field works. (No need to do this for each user role.)

#### Release notes

- Add "Enterprise Fee Summary" report.

Changelog Category: Added

#### Discourse thread

- https://community.openfoodnetwork.org/t/report-that-isolates-various-enterprise-fees/1361

#### Dependencies

- #3142 
- #3088 
- #3213

sauloperez commented on engines/order_management/lib/order_management/reports/enterprise_fee_summary/renderers/html_renderer.rb:
```diff
@@ -0,0 +1,44 @@
+          def header_label(attribute)
+            I18n.t("order_management.reports.enterprise_fee_summary.formats.html.header.#{attribute}")
+          end
+        end
```
This is looking very good but I still fill it's more coupled to the actual report details than about Html, which is the opposite of what we are aiming for.

https://github.com/davidgf/design-patterns-in-ruby/blob/master/strategy.md is exactly what I have in mind. We could name these `enterprise_fee_type_totals` and `list` methods more abstract and with a clear API that makes it a breeze to add a different formatter in the future.

kristinalim replied:
@sauloperez I see what you're saying. Are you okay with pairing the HTML renderer class with a HAML view, but one that is not tied with the controller? For example, a view in `app/views/order_management/reports/enterprise_fee_summary/html_renderer/report.html.haml` of the `order_management` engine. The directory structure would follow `String#underscore` of the renderer.

About using `enterprise_fee_type_totals` and not a more generic method name, those are in anticipation of downloadable reports that contain more than one set of data. For example, a hypothetical report may start with the $$ monthly summary for the year, the next pages to list nearing out of stock items, etc. So what I had in mind was:

1. `ReportService` - Integrates the different classes that are needed to build the report data and render the report
2. `ReportService#enterprise_fee_type_totals` - Represents one table in the report. There can be more!
3. `ReportData::EnterpriseFeeTypeTotals` - This is used by `enterprise_fee_type_totals`. We can make it a convention that table data classes use `#list`. Perhaps we should move the headers in the renderers here?

What do you think?

sauloperez replied:
I don't understand what you mean by, sorry

> The directory structure would follow String#underscore of the renderer.

I guess you mean that we'll pass the renderer object to the haml and it'll reference its methods?

Regarding anticipation, I would not do it. I would not anticipate until we have concrete requirements on the table. In my experience, it leads to YAGNI and the wrong abstractions, which are very costly.

My point was to have the common algorithm that both the CSV and HTML renderers follow implemented in the `Base` class. Now it's almost empty. All renderers should share the same duck-type/interface.

I wanted to share a snippet but I think I went too far :sweat_smile:. Do you fancy a videocall? I think it'll be quicker for both of us.

kristinalim replied:
> The directory structure would follow String#underscore of the renderer.

I just meant that, instead of having the HTML template in `spree/admin/reports/enterprise_fee_summary_report`, `OrderManagement::Reports::EnterpriseFeeSummary::Renderers::HtmlRenderer` owns it so we could have it in (`String#underscore`) `order_management/reports/enterprise_fee_summary/renderers/html_renderer`. But I've changed my mind!

As we discussed in Slack, I realized this HTML template would always be used by the controller, so I'm okay with using the current template path. For any future XLS/PDF/others report generation that will require a view template but ideally can be generated without a controller, we can decide what path to use when that is needed.

kristinalim replied:
> Regarding anticipation, I would not do it. I would not anticipate until we have concrete requirements on the table. In my experience, it leads to YAGNI and the wrong abstractions, which are very costly.
> 
> My point was to have the common algorithm that both the CSV and HTML renderers follow implemented in the Base class. Now it's almost empty. All renderers should share the same duck-type/interface.

I get your point. I will rename `#enterprise_fee_type_totals` to `#data`, and subclass `::EnterpriseFeeTypeTotals` to `::Table` (and probably remove the former altogether!).

kristinalim replied:
For this point, I removed `::EnterpriseFeeTypeTotals` altogether, and just used the method `list` returning an array in the report service. :+1:

sauloperez commented:
I will need a few more sweeps since there's a lot to digest but it's looking really good @kristinalim ! congrats! I just shared few things that I noticed. Let me know what you think.

kristinalim commented on engines/order_management/spec/lib/order_management/reports/enterprise_fee_summary/parameters_spec.rb:
```diff
@@ -0,0 +1,62 @@
+      it { should validate_integer_array(:order_cycle_ids) }
+      it { should validate_integer_array(:enterprise_fee_ids) }
+      it { should validate_integer_array(:shipping_method_ids) }
+      it { should validate_integer_array(:payment_method_ids) }
```
@sauloperez Sorry about that! I never used the `expect` syntax in this form, but this should have been more obvious!

kristinalim replied:
This has been addressed in https://github.com/openfoodfoundation/openfoodnetwork/pull/3115/commits/2628fe02084e06c40d5a829a222188d36ec6d55d. :slightly_smiling_face: Thank you!

sauloperez commented on app/models/spree/payment_method_decorator.rb:
```diff
@@ -23,6 +23,11 @@
 
+  scope :for_distributors, ->(distributors) {
+    non_unique_matches = unscoped.joins(:distributors).where(enterprises: { id: distributors })
+    where(id: non_unique_matches.map(&:id))
```
I don't understand this. In line 28 we execute the query built in line 27, pick up the ids from its result and then querying them again and so getting the same results of line 27 again.

kristinalim replied:
The `joins` here means that you could end up with duplicates. So, this is a workaround to avoid having to use `DISTINCT` or `group` for each query and just have a simple and reusable scope.

```ruby
Spree::PaymentMethod.joins(:distributors).where(enterprises: {id: Enterprise.all}).count
# => 37
Spree::PaymentMethod.count
# => 25
```

I actually have no idea if a subquery will perform better for this, but I guess that the difference is neglible and find this a lot more maintainable.

sauloperez replied:
I was actually talking about the two queries involved here. These are the ones that get generated:

```sql
irb(main):002:0> Spree::PaymentMethod.for_distributors(distributor)
  Spree::PaymentMethod Load (0.8ms)  SELECT "spree_payment_methods".* FROM "spree_payment_methods" INNER JOIN "distributors_payment_methods" ON "distributors_payment_methods"."payment_method_id" = "spree_payment_methods"."id" INNER JOIN "enterprises" ON "enterprises"."id" = "distributors_payment_methods"."distributor_id" WHERE "enterprises"."id" = 2
  Spree::Calculator Load (0.8ms)  SELECT "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = 13 AND "spree_calculators"."calculable_type" = 'Spree::PaymentMethod' LIMIT 1
  Spree::PaymentMethod Load (0.5ms)  SELECT "spree_payment_methods".* FROM "spree_payment_methods" WHERE "spree_payment_methods"."deleted_at" IS NULL AND "spree_payment_methods"."id" IN (13)
  Spree::Calculator Load (0.3ms)  SELECT "spree_calculators".* FROM "spree_calculators" WHERE "spree_calculators"."calculable_id" = 13 AND "spree_calculators"."calculable_type" = 'Spree::PaymentMethod' LIMIT 1
=> [#<Spree::PaymentMethod::Check id: 13, type: "Spree::PaymentMethod::Check", name: "Cheque (Gippsland Poultry)", description: nil, active: true, environment: "development", created_at: "2018-12-11 16:45:07", updated_at: "2018-12-11 16:45:07", deleted_at: nil, display_on: nil>]
```

We fire 4 of them while the first one already does the job. If you look at it, this one loads the payment methods and then we use the ids of the returned rows to fetch them again in the last one. So, one useless roundtrip to the database.

The calculator is something AR loads because it's an association of that payment method (I didn't expect to see it here)

kristinalim replied:
I'm not sure if I'm just missing your point @sauloperez, or possibly also the other way around.

Were you recommending that we should use a join directly, like simply `joins(:distributors).where(enterprises: {id: Enterprise.all})`? My concern is that this could lead to duplicates, like in my example above/below where the result has more than there are total records. It happens when there is a payment method that matches more than one of the enterprises provided.

```ruby
Spree::PaymentMethod.joins(:distributors).where(enterprises: {id: Enterprise.all}).count
# => 37
Spree::PaymentMethod.count
# => 25
```

There is also the option of using `group` or `distinct`, but these would make the named scope not reusable.

So this is one strategy that I think is great, in spite of the extra DB query. Do you disagree?

And yes - it seems even `Spree::PaymentMethod.first` loads the calculator. :sob: I don't want to look into that in this PR.

sauloperez replied:
Ok, I get what you did there 👍 We just had a quick chat and agreed on creating a separate issue to fix it differently. We need to start caring for the DB.

sauloperez commented on engines/order_management/spec/lib/order_management/reports/enterprise_fee_summary/parameters_spec.rb:
```diff
@@ -0,0 +1,62 @@
+    end
+
+    context "for type of parameters" do
+      it { expect(subject).to validate_date_time_format_of(:start_at) }
```
Sorry, I `should` have been more explicit. Pun intended :trollface: 

You have `it { is_expected.to be_empty }` for this. Check it out in https://relishapp.com/rspec/rspec-core/docs/subject/one-liner-syntax

kristinalim replied:
I had no idea this method exists. :blush: Thanks! Fixed in https://github.com/openfoodfoundation/openfoodnetwork/pull/3115/commits/2628fe02084e06c40d5a829a222188d36ec6d55d.

kristinalim commented:
Thanks, @sauloperez, for starting review. :slightly_smiling_face: I will stop rebasing now and will just push commits from now on. Removing WIP from this PR too.

kristinalim commented on config/routes/spree.rb:
```diff
@@ -25,6 +25,7 @@
   match '/admin/reports/products_and_inventory' => 'admin/reports#products_and_inventory', :as => "products_and_inventory_admin_reports",  :via  => [:get, :post]
   match '/admin/reports/customers' => 'admin/reports#customers', :as => "customers_admin_reports",  :via  => [:get, :post]
   match '/admin/reports/xero_invoices' => 'admin/reports#xero_invoices', :as => "xero_invoices_admin_reports",  :via  => [:get, :post]
+  match "/admin/reports/enterprise_fee_summary" => "admin/reports/enterprise_fee_summary_report#index", as: :enterprise_fee_summary_admin_reports,  via: :get
```
I agree! :slightly_smiling_face: Thanks. I have updated this here: https://github.com/openfoodfoundation/openfoodnetwork/pull/3115/commits/e4afeedec95be284dfe37fe87480a3888dfade15#diff-cc818aa9f779759502b5b10331515ef7 I refrained from updating the other lines - I think those should be done in a smaller PR.

kristinalim commented on app/controllers/spree/admin/reports/enterprise_fee_summary_report_controller.rb:
```diff
@@ -0,0 +1,64 @@
+      end
+    end
+  end
+end
```
I actually moved the `lib/order_management/` files into a new engine `order_management`. (See https://github.com/openfoodfoundation/openfoodnetwork/pull/3115/commits/789b30f6faa62e85a0e48799792e09fe77509212 and https://github.com/openfoodfoundation/openfoodnetwork/pull/3115/commits/4f23685ea87341451803c98ca594adf19a6b2655.) But the controller would be more complicated to move :worried: , so I left it in the main application for now. I think we should do that in another PR, as it might need a lot more discussion not related to the Enterprise Fee Summary.

Exposing endpoints from the engine I think would not be tricky after we've packaged core controller behaviours.

As for this file, I have refactored (also because Rubocop demands I do!) and resorted to defining [`report_klass`](https://github.com/kristinalim/openfoodnetwork/blob/feature-enterprise_fees_report/app/controllers/spree/admin/reports/enterprise_fee_summary_report_controller.rb#L45). I think this succeeds in making the class easier to read.

luisramos0 commented on engines/order_management/spec/lib/order_management/reports/enterprise_fee_summary/report_service_spec.rb:
```diff
@@ -0,0 +1,529 @@
+    order = setup_order(options)
+    complete_order(order, options)
+    order.reload
+  end
```
I think this is acceptable but I believe it would be better to use factories for this, like: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/factories.rb#L298

I think it's better because by putting it in factories you would make the code re-usable.

kristinalim replied:
I will do this in a separate PR too, because I'd need to better test the state of the orders when there's a need to label. :slightly_smiling_face: It was trickier to move these methods to a factory, in the first few versions of the file, but it seems doable now.

luisramos0 commented on app/controllers/spree/admin/reports/enterprise_fee_summary_report_controller.rb:
```diff
@@ -0,0 +1,83 @@
+      class EnterpriseFeeSummaryReportController < BaseController
+        before_filter :load_report_parameters, only: [:index]
+        before_filter :load_permissions, only: [:index]
+        before_filter :load_authorizer, only: [:index]
```
I agree.

kristinalim replied:
This is no longer a problem in the current version. :+1:

luisramos0 commented on config/locales/en.yml:
```diff
@@ -2643,6 +2662,44 @@ See the %{link} to find out more about %{sitename}'s features and to start using
+              tax_category_name: "Tax Category"
+              total_amount: "$$ SUM"
+        invalid_filter_parameters: "The filters you selected for this report are invalid."
+
```
agree but... one step at a time :-)

sauloperez replied:
I just say so because Rails engines are almost regular Rails apps and they also have a `config/locale/`

> or locales, simply place the locale files in the config/locales directory, just like you would in an application.

kristinalim replied:
@sauloperez Yes, it would be nice, but we would still have to see how to work the Transifex integration for that.

We can start by moving the translations needed for the `order_management` engine to the `order_management` and `js.order_management` I18n scopes, though. I will do this.

luisramos0 replied:
oh, interesting. we need to make lazyloading work inside engines.
for example, will t('.new_order') in engines/order_management/app/views/order_management/orders/new.html.haml work as  order_management.orders.new.new_order = "New Order" in en.yml?

kristinalim replied:
I realized that only the translations for the JS, views, and controller are not yet in the "order_management" I18n scope, but I don't plan to move those in the engine yet because that would be quite complicated. I will just create a tech debt issue for moving these to the engine.

Translations for the services are already in "order_management", so I'm marking this resolved already. :slightly_smiling_face: But if you have objections, feel free to undo.

kristinalim commented on app/controllers/spree/admin/reports/enterprise_fee_summary_report_controller.rb:
```diff
@@ -0,0 +1,64 @@
+          @permissions = OrderManagement::Reports::EnterpriseFeeSummary::Permissions.new(spree_current_user)
+          @authorizer = OrderManagement::Reports::EnterpriseFeeSummary::Authorizer.new(@report_parameters, @permissions)
+
+          return render_report_form if params[:report].blank?
```
`#new` and `#create` sounds good to me, @sauloperez! Moving report generation to `#create` even makes sense if we'll eventually have to make this asynchronous for performance reasons.

If we're separating these into two actions, maybe it's time to change this into a resource. I can define this in the engine too, without the 'order_management` namespace:

```ruby
namespace :admin do
  namespace :reports do
    resource :enterprise_fee_summary, only: [:new, :create]
  end
end
```

I would rename the controller from `::Reports::EnterpriseFeeSummary**Controller**` to `::Reports::EnterpriseFeeSummaryController`.

What do you think?

sauloperez replied:
Yes please a resource would be awesome! When I implemented our katuma_reports it was the best way I found to implement a report. I google what the community says about it and what apps like Github do. You can check it out in https://github.com/coopdevs/katuma_reports/blob/master/config/routes.rb. Beware, there's :hankey: that is totally unrelated :see_no_evil:

sauloperez commented:
@kristinalim this is awesome! I just shared few ideas I think can make this even better. Let's see if they fit here or on an upcoming PR.

sauloperez commented on engines/order_management/lib/order_management/reports/enterprise_fee_summary/scope.rb:
```diff
@@ -0,0 +1,361 @@
+      end
+    end
+  end
+end
```
:metal: SQL! I think though that by making use AR's DSL we can make this a bit more maintainable. On the other hand, I feel like we just put together all these queries into a class because the just touch the same tables, leading to poor cohesion. It :nose: too attached to AR scopes.

When I have to implement a rather complex query I find the query object pattern very handy. I first read about it in the famous Code Climate's post https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/ and the folks at Thoughtbot have a nice practical case as well in https://robots.thoughtbot.com/active-record-eager-loading-with-query-objects-and-decorators#the-query-object

luisramos0 replied:
nice! I studied this pattern in my MsC back in 2004 when "P of EAA" was a new book!!!
https://martinfowler.com/eaaCatalog/queryObject.html
In this link you find a more advanced case where the query object can be composed of variations/subclasses/query_criteria, I dont think we need this here but I am happy to share some history :-D

kristinalim replied:
I did not use named scopes here because Rails 3 doesn't allow use of named scopes for left joins, many of the joins need custom join names (for example, `enterprises` and `exchanges` tables are used more than once), and there are many `ON` conditions that presume certain joins are already available. These makes it complicated to ensure that scope chaining would still work with other existing named scopes.

I will see if there is a reliable gem for Rails 3 that would allow us to use left join for named scopes, and if yes, see what is reusable enough and can be moved to a model. :slightly_smiling_face:

kristinalim replied:
Thank you for the resources, @sauloperez and @luisramos0! :slightly_smiling_face: So much to learn, and motivated to study more so I can share some back too.

But for now, sharing that I was mind-blown :heart_eyes: when I realized how the very simple the concept of returning `self` is how `ActiveRecord::Relation` allows method chaining:

```ruby
def chain_to_scope(&block)
  @scope = @scope.instance_eval(&block)
  self
end
```

Hehe, it might be obvious, but I didn't really think about it before. Used method chaining in an initial version of the Scope class.

luisramos0 replied:
yeah, chaining is commonly done by returning self ("return this" in java and javascript). And hopefully it's a form of the builder pattern, otherwise it's probably just someone trying to be smart :-)

sauloperez replied:
my point was more about having bits and pieces of AR where we can so things are a bit more maintainable. Of course there's no support for LEFT JOIN but we can have just these as SQL or [Arel](https://jpospisil.com/2014/06/16/the-definitive-guide-to-arel-the-sql-manager-for-ruby.html) if it makes sense.

kristinalim replied:
Thanks @sauloperez! Okay, I will revisit. I've never used `Arel` directly yet.

kristinalim replied:
I will create a tech debt issue for investigating reusing some model associations here or using Arel. I'm marking this review thread resolved already. :slightly_smiling_face: But if you have objections, feel free to undo.

kristinalim commented on app/controllers/spree/admin/reports/enterprise_fee_summary_report_controller.rb:
```diff
@@ -0,0 +1,64 @@
+          @authorizer = OrderManagement::Reports::EnterpriseFeeSummary::Authorizer.new(@report_parameters, @permissions)
+
+          return render_report_form if params[:report].blank?
+          return respond_to_invalid_parameters unless @report_parameters.valid?
```
The current form of this method is:

```ruby
def index
  return render_report_form if params[:report].blank?
  return respond_to_invalid_parameters unless @report_parameters.valid?

  @authorizer.authorize!
  @report = report_klass::ReportService.new(@permissions, @report_parameters, report_renderer_klass)
  render_report
rescue OpenFoodNetwork::Reports::Authorizer::ParameterNotAllowedError => e
  flash[:error] = e.message
  render_report_form
end
```

The line `return render_report_form if params[:report].blank?` would no longer be necessary if we have a dedicated `#new` action.

Regaring the form of the `if ... else` block, our Rubocop does prefer guard clauses though...

I read your comment about using controller filters for single actions. With `#new` and `#create`, we will have two actions now. I can kill the `before_filter` for `@authorizer` and set it up in the `#index` method.

```ruby
def create
  return respond_to_invalid_parameters unless @report_parameters.valid?
  authorize_for_report!(@report_parameters, @permissions)
  @report = report_klass::ReportService.new(@permissions, @report_parameters, report_renderer_klass)
  render_report
rescue OpenFoodNetwork::Reports::Authorizer::ParameterNotAllowedError => e
  flash[:error] = e.message
  render_report_form
end

private

def authorize_for_report!(report_parameters, permissions)
  authorizer = report_klass::Authorizer.new(report_parameters, permissions)
  authorizer.authorize!
end
```

I'm very tempted to define `authorize!(permissions)` in `::Parameters`. `::Authorizer` is a parameter authorizer after all, and permission logic is in `::Permissions`.

What do you think?

sauloperez replied:
Sounds good to me

kristinalim commented on lib/open_food_network/reports.rb:
```diff
@@ -0,0 +1,5 @@
+module OpenFoodNetwork
```
In the discussion about engines, we talked about classes for conventions among domain-engines which we could put in a gem or engine. I prefer engine and not gem, because of the bigger overhead in maintaining a gem. Perhaps we should already move this to that new engine? We can do that in a separate PR.

When we move forward and want to separate each domain-engine into a different application, we can extract this engine into a gem, and each domain-engine application can upgrade the gem at its own pace.

luisramos0 replied:
first, guys, it feels so good we are having this conversation!!!

I agree with the reasoning about creating a common engine and not a gem as long as this engine does not have models in it, just services. we can call this engine "commons" :-)

But as general YAGNI, I don't think we should abstract things into that engine until we have 2 engines needing it. We may end up having all reports in this order_management engine (most of them belong here right?) or even end up moving all reports to a separate engine of it's own...
so, I am ok with (1) creating that commons engine now with common reports stuff in it, but for now (2) I'd just move these things into the orders_management engine /app/services/reports folder. your choice (what is important is to stop using open_food_network namespace).

sauloperez replied:
Agree with @luisramos0 on 2

kristinalim replied:
Sounds great, @luisramos0. :-) I agree too.

kristinalim commented on engines/order_management/lib/order_management/reports/enterprise_fee_summary/authorizer.rb:
```diff
@@ -0,0 +1,40 @@
+
+          raise error_klass, error_message unless ids_allowed
+        end
+      end
```
@sauloperez Oh! :grinning: Indeed, we can use CanCan for that! I knew I didn't want to put the more elaborate authorization logic in the `Ability`, but it didn't occur to me to pass the parameter and authorizer objects there.

I'll look into how to do it. IIRC, there's a way to pass additional context to the `Ability`.

kristinalim replied:
I haven't used `Pundit` yet, but there is `CanCanCan` which is maintained, closer or almost identical to `CanCan`, and [used by Spree `master`](https://github.com/spree/spree/blob/master/core/spree_core.gemspec#L27). So I think we should stick to `CanCan` for now.

sauloperez replied:
Yes, let's continue with Cancancan for this engine :+1: My point was not to implement yet another authorization library.

kristinalim replied:
I think there will still be some discussion needed for implementation of bringing authorization to `CanCan`, so let's tackle this in a separate issue too.

kirstenalarsen commented:
Hi everyone - hard to follow the conversation above - are we close or not? PLEASE can we get this into staging to test before xmas?

kirstenalarsen commented:
hello? @kristinalim @sauloperez @luisramos0 - perfect is enemy of good / done?! Can we please get this into staging for testing asap?!

kristinalim commented:
@kirstenalarsen I will finish addressing the easier changes in the review comments by tomorrow. Then, @sauloperez @luisramos0 perhaps we can tackle the other changes proposed as follow-up changes?

sauloperez commented:
I understand your sense of hurry here but it doesn't do any good. We still have to improve how we iteratively ship new implementations. As we know, bigger PRs take longer to review and this is one in particular affects a known tricky part of the codebase.

> PLEASE can we get this into staging to test before xmas?

I didn't know we had a set deadline. We could have organized differently.

> @sauloperez @luisramos0 perhaps we can tackle the other changes proposed as follow-up changes?

I'm sorry it's a bad week for me and I can't help.

kristinalim commented:
> @sauloperez @luisramos0 perhaps we can tackle the other changes proposed as follow-up changes?

What I was proposing is that we can settle with whatever presumably working implementation I can finish until tomorrow, if there are no bugs identified in code review get that to testing phase, and then potentially live. And we can just tackle the other suggested changes as "tech debt" which will be done immediately too.

Are you okay with that?

sauloperez commented:
Absolutely! this is looking very good already!

luisramos0 commented:
My only point is to move the files from /lib to app/services. Are you guys saying you don't want to do it now? I am ok with it as long an issue is opened for it and that issue gets done soon afterwards.

kristinalim commented:
@daniellemoorhead Yes, I've already rebased locally. Still working on a couple of the easier changes from their review, and pushing everything here in a bit. The Code Climate check is failing because of an issue that's already fixed in master.

@luisramos0 That will be in `app/services/` of the engine already. :+1:

kristinalim commented on engines/order_management/lib/order_management/reports/enterprise_fee_summary/report_data/enterprise_fee_type_totals.rb:
```diff
@@ -0,0 +1,19 @@
+      end
+    end
+  end
+end
```
I removed `::EnterpriseFeeTypeTotals` altogether, and just used the method `list` returning an array in the report service. :+1:

kristinalim commented on engines/order_management/lib/order_management/reports/enterprise_fee_summary/authorizer.rb:
```diff
@@ -0,0 +1,40 @@
+
+module OrderManagement
+  module Reports
+    module EnterpriseFeeSummary
```
Thanks, @luisramos0! This has been done.

luisramos0 replied:
awesome, thank you!

kristinalim commented:
This is ready for another round of reviews. :joy: Moving this back to Code Review.

I created tech debt issues for other suggested changes that can follow:

* #3214 
* #3215 
* #3216 
* #3217

kristinalim commented:
Unless there are other changes needed before this can get tested, I would like to continue working on #3085. RIght now, tests in engines aren't getting run. I would like to have this working properly before this PR can be merged.

`bundle exec rspec engines/order_management/spec` passes when I run it.

kristinalim commented:
> This PR is 51 commits and 3003 new lines, it's way too big. Maybe we should have an action here to discuss among ourselves the strategies to avoid this situation in the future.

Oh yikes. Yes, it is indeed too big @luisramos0. I wondered about creating one base PR on which to merge other smaller PRs, but I did not push through with it because I didn't think it would get this big. I should have done that.

luisramos0 commented:
@sauloperez second approver moves to test ready ;-)

sauloperez commented:
Ups, sorry

mkllnk commented:
Staged on https://staging.openfoodnetwork.org.au/.

kristinalim commented:
As discussed [on Slack](https://openfoodnetwork.slack.com/messages/G44L6NRQU/convo/C2GQ45KNU-1547186925.158100/), let's merge this approved PR to the new transitional branch `feature/enterprise_fee_summary`, which will hold the changes until everything is ready to be tested one last time and merged into master when good. This is so that subsequent changes (including fixes) can be separate from this already big PR, and be reviewed, tested, and merged incrementally.

I've already made changes so that issues connected to this PR are not automatically closed when this is merged. The issues can be closed manually after receiving approval from the testers.

* Updated the base branch to the new transitional branch, not `master`.
* Updated the issue description to not close related issues automatically.
* Disconnected this PR from connected issues on Zenhub, so they do not have to move through the pipe together.
